File: test_expected_contribution_simple.js
// Simple test script to check expected contribution calculation
const { createClient } = require('@supabase/supabase-js');
require('dotenv').config();

const supabaseUrl = process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY;
const supabase = createClient(supabaseUrl, supabaseKey);

async function testExpectedContribution() {
  console.log('Testing expected contribution calculation...\n');
  
  // Test with a few member numbers
  const testMembers = ['M004', 'M005', 'M031', 'M047', 'M057'];
  
  for (const memberNumber of testMembers) {
    console.log(`\n=== Testing member ${memberNumber} ===`);
    
    try {
      // Get member data
      const { data: member, error } = await supabase
        .from('members')
        .select('*')
        .eq('member_number', memberNumber)
        .single();
      
      if (error) {
        console.log(`❌ Error fetching member: ${error.message}`);
        continue;
      }
      
      if (member) {
        console.log(`Name: ${member.name}`);
        console.log(`Member Number: ${member.member_number}`);
        console.log(`Join Date: ${member.join_date}`);
        
        // Parse financial_info
        let financialInfo = {};
        if (member.financial_info) {
          try {
            financialInfo = typeof member.financial_info === 'string' 
              ? JSON.parse(member.financial_info) 
              : member.financial_info;
          } catch (e) {
            console.log(`Warning: Could not parse financial_info`);
          }
        }
        
        console.log(`Current Expected Contribution in DB: R${financialInfo.expected_contribution || 0}`);
        
        // Calculate expected contribution based on join date
        if (member.join_date) {
          const joinDate = new Date(member.join_date);
          const today = new Date();
          
          // Calculate months difference
          const monthsDiff = (today.getFullYear() - joinDate.getFullYear()) * 12 + 
                            (today.getMonth() - joinDate.getMonth());
          
          // Cap at 83 months (maximum expected contribution period)
          const cappedMonths = Math.max(0, Math.min(monthsDiff, 83));
          
          // Calculate expected contribution (R200 per month)
          const expectedContribution = cappedMonths * 200;
          
          console.log(`Join Date: ${joinDate.toISOString().split('T')[0]}`);
          console.log(`Months since join: ${monthsDiff}`);
          console.log(`Capped months (max 83): ${cappedMonths}`);
          console.log(`Calculated Expected Contribution: R${expectedContribution}`);
          
          if (financialInfo.expected_contribution !== expectedContribution) {
            console.log(`⚠️ WARNING: Expected contribution mismatch!`);
            console.log(`   Database: R${financialInfo.expected_contribution || 0}`);
            console.log(`   Calculated: R${expectedContribution}`);
            
            // Update the expected contribution in the database
            const updatedFinancialInfo = {
              ...financialInfo,
              expected_contribution: expectedContribution,
              expected_contribution_details: {
                total_months: cappedMonths,
                period1_months: Math.min(cappedMonths, 60), // First 5 years
                period2_months: Math.max(0, cappedMonths - 60) // After 5 years
              },
              last_expected_contribution_update: new Date().toISOString()
            };
            
            const { error: updateError } = await supabase
              .from('members')
              .update({
                financial_info: updatedFinancialInfo,
                last_updated: new Date().toISOString()
              })
              .eq('member_number', memberNumber);
            
            if (updateError) {
              console.log(`❌ Error updating expected contribution: ${updateError.message}`);
            } else {
              console.log(`✅ Updated expected contribution to R${expectedContribution}`);
            }
          } else {
            console.log(`✅ Expected contribution matches calculation`);
          }
        } else {
          console.log(`⚠️ No join date found for member`);
        }
      } else {
        console.log(`❌ Member ${memberNumber} not found`);
      }
    } catch (error) {
      console.error(`Error testing member ${memberNumber}:`, error.message);
    }
  }
  
  console.log('\n=== Testing complete ===');
}

// Run the test
testExpectedContribution().catch(console.error);

File: test_expected_contribution_simple.py
import os
from supabase import create_client
from dotenv import load_dotenv
import json

load_dotenv()
SUPABASE_URL = os.getenv('SUPABASE_URL')
SUPABASE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
supabase = create_client(SUPABASE_URL, SUPABASE_KEY)

def test_expected_contribution():
    print('Testing expected contribution in database...\n')
    
    # Test with a few member numbers
    test_members = ['M004', 'M005', 'M031', 'M047', 'M057']
    
    for member_number in test_members:
        print(f'\n=== Testing member {member_number} ===')
        
        try:
            # Get member data
            response = supabase.table('members').select('*').eq('member_number', member_number).single().execute()
            member = response.data
            
            if member:
                # Parse financial_info
                financial_info = member.get('financial_info', {})
                if isinstance(financial_info, str):
                    try:
                        financial_info = json.loads(financial_info)
                    except:
                        financial_info = {}
                
                name = member.get('name', 'Unknown')
                join_date = member.get('join_date', '2018-07-23')
                expected_contribution = financial_info.get('expected_contribution', 0)
                
                print(f'Name: {name}')
                print(f'Member Number: {member_number}')
                print(f'Join Date: {join_date}')
                print(f'Expected Contribution in DB: R{expected_contribution}')
                
                # Calculate expected contribution based on join date
                from datetime import datetime
                join_date_obj = datetime.fromisoformat(join_date.replace('Z', '+00:00'))
                today = datetime.now()
                months_diff = (today.year - join_date_obj.year) * 12 + (today.month - join_date_obj.month)
                calculated_expected = max(0, min(months_diff, 83)) * 200
                
                print(f'Months since join: {months_diff}')
                print(f'Calculated Expected Contribution: R{calculated_expected}')
                
                if expected_contribution != calculated_expected:
                    print(f'⚠️ WARNING: Expected contribution mismatch!')
                    print(f'   Database: R{expected_contribution}')
                    print(f'   Calculated: R{calculated_expected}')
                else:
                    print(f'✅ Expected contribution matches calculation')
            else:
                print(f'❌ Member {member_number} not found')
        except Exception as e:
            print(f'Error testing member {member_number}: {str(e)}')
    
    print('\n=== Testing complete ===')

if __name__ == '__main__':
    test_expected_contribution()